libdevcore/termcolor.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libdevcore/termcolor.h>
+
+#include <iostream>
+#include <ostream>
+#include <utility>
+
+#if defined(__unix__)
+#include <unistd.h>
+#endif
+
+#if defined(_WIN32)
+#include <windows.h>
+#endif
+
+using namespace std;
+
+namespace termcolor {
+
+namespace internal {
+
+static int const enabled_index = std::ios_base::xalloc();
+
+void setupTerminal()
+{
+#if defined(_WIN32) && defined(ENABLE_VIRTUAL_TERMINAL_PROCESSING)
+	// Set output mode to handle virtual terminal (ANSI escape sequences)
+	// ignore any error, as this is just a "nice-to-have"
+	// only windows needs to be taken care of, as other platforms (Linux/OSX) support them natively.
+	HANDLE hOut = GetStdHandle(STD_OUTPUT_HANDLE);
+	if (hOut == INVALID_HANDLE_VALUE)
+		return;
+
+	DWORD dwMode = 0;
+	if (!GetConsoleMode(hOut, &dwMode))
+		return;
+
+	dwMode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
+	if (!SetConsoleMode(hOut, dwMode))
+		return;
+#endif
+}
+
+}  // namespace internal
+
+bool is_terminal(ostream const& os)
+{
+	if (&os == &cout)
+#if defined(__unix__)
+		return isatty(STDOUT_FILENO);
+#else
+		return true;
+#endif
+
+	if (&os == &cerr)
+#if defined(__unix__)
+		return isatty(STDERR_FILENO);
+#else
+		return true;
+#endif
+
+	return false;
+}
+
+bool is_enabled(ostream const& os)
+{
+	return const_cast<ostream&>(os).iword(internal::enabled_index) != 0;
+}
+
+ostream& enable(ostream& os)
+{
+	if (!is_terminal(os))
```

Comment:
Shouldn't this read `if (is_terminal` instead?

Reply:
No, because while stdout (cout) can be redirected by the users, it doesn't mean the user has redirected stderr (cerr), for example.

Just like the libc call `isatty(fd)` requires a file descriptor to check, too).